### DISCUSSION TITLE
Rendering Gradient in Multiple Passes - dr.grad Seed?

### DISCUSSION THREAD

**Discussion by coult099** [mitsuba-renderer/mitsuba3#1330] [Q&A] [answered]
Hello,

I have a rendering job that requires relatively noise-free images and gradients. In order to render a 256 x 512 image with relatively high sample count (> 2^18 SPP), I have been rendering in multiple passes (using different seeds for each pass) & averaging the result. Is there a similar way to render an auto-diff gradient in multiple passes? I have included code for what I think it should be, and would appreciate guidance here. Thank you.

```python
import mitsuba as mi
import drjit as dr

scene = mi.load_file('./cbox/cbox.xml')
params = mi.traverse(scene)
key = 'green.reflectance.value'

single_spp = 2**14
nRenders = 8

img_tmp = dr.zeros(mi.TensorXf, (res_h, res_w, 3))
grd_tmp = dr.zeros(mi.TensorXf, (res_h, res_w, 3))

for seed_num in range(nRenders):
    img_tmp += mi.render(scene, params, spp=single_spp, seed=seed_num)
    dr.forward(params[key])
    grd_tmp += dr.grad(img_tmp)

image = img_tmp / nRenders
grad_image = grd_tmp / nRenders

mi.util.write_bitmap("image.png", image)
mi.util.write_bitmap("gradient.png", grad_image)
```

**Comment by merlinND** [comment, marked as answer]
Hello @coult099,

Have you tried the `seed_grad` argument of `mi.render()`?

**Reply by coult099** [reply]
This is what I was looking for, thank you!

**Reply by coult099** [reply]
Hello, I tried setting the `seed_grad` argument differently on different passes (and used a different value from the `seed` argument), but I am seeing similar/the same noise characteristics in successive passes of the gradient computation with different seeds. Do you have any idea why this might be?

**Reply by merlinND** [reply]
If you use a lower spp and make sure to use the `independent` sampler, do you see the noise patterns in the gradients change with each `seed_grad` ?
